DR-2935 Add upgrade method to convert random file ids to predicable file ids

nmalfroy opened the pull request:
Adds an upgrade utility method that will update the file ids from random ids to predictable ids by:
- Obtaining the existing ids
- Determining the new ones
- Storing the difference in Firestore
- Updating the BigQuery data with pointers to files by:
  - soft deleting existing rows (for tables with at least one file or directory ref)
  - adding new rows with new ids (for tables with at least one file or directory ref)
 

All while enforcing that no snapshot must exist on the dataset

Note: most testing was done manually though connected tests added to verify the trickier pieces.

To call the upgrade, you invoke:
```
curl -X 'POST' \
  'https://<hostname>/api/repository/v1/upgrade' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer xxx' \
  -H 'Content-Type: application/json' \
  -d '{
  "upgradeName": "CONVERT_DATASET_FILE_IDS",
  "upgradeType": "custom",
  "customName": "CONVERT_DATASET_FILE_IDS",
  "customArgs": [
    "<your dataset id>"
  ]
}'
```

nmalfroy committed:
Move firestore querying method to be reused and add new method to scan (bulk) delete

nmalfroy committed:
Add option to filter files to only files when retrieving fileid from firestore

(specifically, ignoring directories)

nmalfroy committed:
Add methods to update the predictable ids flag

nmalfroy committed:
Add BigQueryPDao methods to handle file id updates

nmalfroy committed:
Add the flight that will upgrade a single dataset

nmalfroy committed:
Register upgrade and add support for passing in flight inputs

snf2ye commented on src/main/java/bio/terra/service/tabulardata/google/bigquery/BigQueryDatasetPdao.java:
```diff
@@ -565,6 +629,64 @@ public void insertIntoDatasetTable(
+  private static final String insertNewFileIdsIntoDatasetTableTemplate =
+      """
+    /* Soft delete all existing rows */
+    INSERT INTO `<project>.<dataset>.<softDeleteTable>` (<transactIdColumn>, <rowIdColumn>, <flightIdColumn>, <deletedAtColumn>, <deletedByColumn>)
+    SELECT @transactId, <rowIdColumn>, @flightId, CURRENT_TIMESTAMP(), @deletedBy
+    FROM `<project>.<dataset>.<liveTable>`;
+
+    /* Insert rows with new values */
+    INSERT INTO `<project>.<dataset>.<targetTable>` (<transactIdColumn>, <rowIdColumn>, <columns:{c|<c.name>}; separator=",">)
+    SELECT @transactId,GENERATE_UUID(),<columns:{c|<if(c.fileOrDirRef)>
+      <if(c.arrayOf)>
+        ARRAY(SELECT f.<newIdColumn> FROM UNNEST(<c.name>) i JOIN `<project>.<dataset>.<fileIdMappingTable>` f ON i = f.<origIdColumn>)
+      <else>
+        (SELECT <newIdColumn> FROM `<project>.<dataset>.<fileIdMappingTable>` f WHERE f.<origIdColumn> = t.<c.name>)
+      <endif>
+    <else>
+      <c.name>
+    <endif>}; separator=",">
+    FROM `<project>.<dataset>.<liveTable>` t;
+      """;
```
I'm surprised to see these one-off queries - Why aren't we using the existing update mechanism? (Which, should do this under the covers - soft delete & insert)

nmalfroy replied:
I think because this is a special usecase where we need to join with the new table.  It's pretty low level, which is where we'd normally see these types of one offs that get hidden in the service layer.  I suppose we could have put this into the merge machinery but that seemed like it could complicate things

snf2ye replied:
My main concern is around maintainability long term -- If we change how transactions or our soft delete set up works, we'll need to remember to also update this code.

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java:
```diff
@@ -143,6 +149,48 @@ <V> void scanCollectionObjects(
+              " scanning " + batchSize + " items for collection id: " + collectionId);
+      batchCount++;
+      if (!documents.isEmpty()) {
+        logger.info("Visiting batch " + batchCount + " of ~" + batchSize + " documents");
```
There's no way to easily obtain how many total documents we're processing, right? Just have to fetch in batches until we have none left? (Or maybe the total is logged earlier?)

I ask because I find logs like this the most helpful when they clearly indicate your progress towards a goal.

nmalfroy replied:
Yeah I know, it stinks but there's no way to get the full count without making another full scan query :(

okotsopoulos replied:
No problem, I figured that was the case!

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java:
```diff
@@ -248,7 +296,7 @@ <T, V> List<T> batchOperation(List<V> inputs, ApiFutureGenerator<T, V> generator
       // Exponential backoff
       logger.info(
-          "[batchOperation] will attempt retry #{} after {} millisecond pause. {} requests completed this round.",
+          "[batchOperation] will attempt retry #{} after {} second pause. {} requests completed this round.",
```
Good catch!

okotsopoulos commented on src/test/java/bio/terra/service/dataset/DatasetDaoTest.java:
```diff
@@ -904,6 +904,19 @@ public void createDatasetWithProperties() throws Exception {
+        datasetDao.retrieve(datasetId).hasPredictableFileIds());
+    datasetDao.setPredictableFileId(datasetId, true);
+    assertTrue(
+        "predictable file ids flag is false",
```
I think this should be "predictable file ids flag is true" :)

nmalfroy replied:
sigh, you are right...will update

okotsopoulos commented on src/test/java/bio/terra/service/dataset/DatasetDaoTest.java:
```diff
@@ -904,6 +904,19 @@ public void createDatasetWithProperties() throws Exception {
+    assertTrue(
+        "predictable file ids flag is false",
+        datasetDao.retrieve(datasetId).hasPredictableFileIds());
+    datasetDao.delete(datasetId, TEST_USER);
```
Is there a reason that we delete at the end of the test rather than in an `@After` teardown method? (I see we do this in other tests in this class too)

nmalfroy replied:
Yeah, this is an old test and I wanted to just follow convention (especially since some of the tests do a bunch of deleting as part of their tests.  I can update it though.  Lol, gonna my my line counts go up :)

okotsopoulos commented on src/test/java/bio/terra/service/dataset/DatasetServiceUnitTest.java:
```diff
@@ -110,4 +114,15 @@ public void patchDatasetIamActions() {
+    when(summary.toModel()).thenReturn(new DatasetSummaryModel().id(datasetId));
+    when(datasetDao.retrieveSummaryById(datasetId)).thenReturn(summary);
+    datasetService.setPredictableFileIds(datasetId, true);
+    verify(datasetDao, times(1)).setPredictableFileId(eq(datasetId), eq(true));
+    verify(datasetDao, times(1)).retrieveSummaryById(eq(datasetId));
```
If you wish, these lines could be consolidated thusly:

```
verify(datasetDao).setPredictableFileId(datasetId, true);
verify(datasetDao).retrieveSummaryById(datasetId);
```
(`verify` defaults to verifying a single occurrence and you don't need the `eq` matchers if you're checking all specific args)

nmalfroy replied:
oh cool.  TIL! Will update

okotsopoulos commented on src/main/java/bio/terra/service/tabulardata/google/bigquery/BigQueryDatasetPdao.java:
```diff
@@ -482,6 +485,67 @@ private static BulkLoadHistoryModel bigQueryResultToBulkLoadHistoryModel(
+    List<FileIdMapping> fileIdArray =
+        oldToNewMappings.entrySet().stream()
+            .map(e -> new FileIdMapping(e.getKey(), e.getValue()))
+            .collect(Collectors.toList());
```
You could consolidate this to `.toList()` if you'd like.

nmalfroy replied:
Gah, right you are.  Will update

okotsopoulos commented on src/main/java/bio/terra/service/tabulardata/google/bigquery/BigQueryDatasetPdao.java:
```diff
@@ -565,6 +629,64 @@ public void insertIntoDatasetTable(
 
+  private static final String insertNewFileIdsIntoDatasetTableTemplate =
+      """
+    /* Soft delete all existing rows */
```
I really appreciate the comments in this statement!

okotsopoulos commented on src/main/java/bio/terra/service/tabulardata/google/bigquery/BigQueryDatasetPdao.java:
```diff
@@ -565,6 +629,64 @@ public void insertIntoDatasetTable(
+      throws InterruptedException {
+    BigQueryProject bigQueryProject = BigQueryProject.from(dataset);
+
+    ST sqlTemplate = new ST(insertNewFileIdsIntoDatasetTableTemplate);
```
Personal preference but I've been finding it more readable to chain the ST parameter specs, like so:
```
ST sqlTemplate = new ST(insertNewFileIdsIntoDatasetTableTemplate)
  .add("project", bigQueryProject.getProjectId())
  .add("dataset", BigQueryPdao.prefixName(dataset.getName()))
 ….
```

nmalfroy replied:
ah, that's a great call.  I'll change to the fluent style

nmalfroy replied:
I'll actually do it for all of the `BigQuery*Pdao` classes (multiline editing FTW!)

okotsopoulos commented on src/test/java/bio/terra/service/tabulardata/google/BigQueryPdaoTest.java:
```diff
@@ -372,6 +381,110 @@ public void testGetSnapshotTableData() throws Exception {
+      bigQueryDatasetPdao.deleteDataset(dataset);
+      assertThatDatasetAndTablesShouldExist(dataset, false);
+    } finally {
+      datasetDao.delete(dataset.getId(), TEST_USER);
```
I'm sorry, I think I've asked about this in previous PRs, but is there a reason that we don't delegate deletion to the `@After` teardown method?

nmalfroy replied:
legacy tests...I started going through this one and it's a bit of a weird one (it partially creates datasets, I think to speed up the test.  I'm gonna leave this one alone for now since it's gonna be a major refactor of this class

okotsopoulos replied:
No problem! Makes sense to cal that OOS

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertFileIdUtils.java:
```diff
@@ -0,0 +1,23 @@
+
+public class ConvertFileIdUtils {
+
+  private ConvertFileIdUtils() {}
```
Curious, do you actually need the no-op constructor?

nmalfroy replied:
Yeah...the linter complains without it.  It enforces that you can't instantiate this class

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsGetIdsStep.java:
```diff
@@ -0,0 +1,75 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ConvertToPredictableFileIdsGetIdsStep implements Step {
```
You could extend `DefaultUndoStep` instead since `undoStep` is a no-op.

nmalfroy replied:
Ah right!

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsBqCreateStageTableStep.java:
```diff
@@ -0,0 +1,64 @@
+        ConvertFileIdUtils.readFlightMappings(context.getWorkingMap());
+
+    if (oldToNewMappings.isEmpty()) {
+      logger.info("No file ids to migrate");
```
For my own understanding: we'd only get to this point if the dataset tables contain no file IDs.  If a dataset has previously been converted to use predictable file IDs, we get a nonempty `oldToNewMappings` map, right?

nmalfroy replied:
that's exactly right.  I'll add a code comment to that effect

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsBqCreateStageTableStep.java:
```diff
@@ -0,0 +1,64 @@
+    Map<UUID, UUID> oldToNewMappings =
+        ConvertFileIdUtils.readFlightMappings(context.getWorkingMap());
+
+    if (oldToNewMappings.isEmpty()) {
```
Looks like we're only using `oldToNewMappings` to check for its size.  Not sure how resource-intensive it is to run `readFlightMappings` to perform the conversion, but if we're concerned about that we could also store the map size in the working map rather than having to run the conversion several times.

nmalfroy replied:
Interesting.  FWIW, I've been testing with multi MB objects as part of bulk ingest work and it's subsecond.  I might leave it alone just to make it less things that need to be tracked

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsBqStageDataStep.java:
```diff
@@ -0,0 +1,50 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ConvertToPredictableFileIdsBqStageDataStep implements Step {
```
Another one that could extend `DefaultUndoStep` :) (I'll stop making this comment, you get the idea!)

nmalfroy replied:
haha yep, I went and updated the ones that could be updated

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsBqUpdateRowsStep.java:
```diff
@@ -0,0 +1,67 @@
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    Dataset dataset = datasetService.retrieve(datasetId);
```
I see that we retrieve the dataset anew in many steps.  Would it make sense to retrieve it earlier in the flight and pass that around instead?

nmalfroy replied:
I was wondering about that but honestly, it works out to roughly the same: you're reading from the DB in all cases

okotsopoulos replied:
Fair enough!

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsBqUpdateRowsStep.java:
```diff
@@ -0,0 +1,67 @@
+    Map<UUID, UUID> oldToNewMappings =
+        ConvertFileIdUtils.readFlightMappings(context.getWorkingMap());
+
+    if (oldToNewMappings.isEmpty()) {
```
Similar comment here around how we use `oldToNewMappings` only to check its size.

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsUpdateMetadataStep.java:
```diff
@@ -0,0 +1,26 @@
+import bio.terra.stairway.exception.RetryException;
+import java.util.UUID;
+
+public class ConvertToPredictableFileIdsUpdateMetadataStep extends DefaultUndoStep {
```
Actually, don't we want to reset the metadata value in `undoStep` if we get there?

nmalfroy replied:
Aha, it actually doesn't matter since it runs in a transaction.  Though I'm realizing I didn't set the rollback option to true in the flight.  Will do that

nmalfroy replied:
Er rather, it doesn't roll back so it just doesn't show up.  Either way, we should be ok

okotsopoulos replied:
Ohhhh!  Okay, sounds good!

When I see a step that doesn't clean up after itself because the clean up is handled in some way elsewhere in the flight, I find it helpful to have a comment to that effect. If you wanted to unit test this, I'd check the flight construction to make sure that this step is scheduled after a transaction opened step.  But I won't hold up my approval, these are thoughts more applicable to enduring code.

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsBqUpdateRowsStep.java:
```diff
@@ -0,0 +1,67 @@
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    return StepResult.getStepResultSuccess();
```
Do we need to revert the file IDs loaded in an `undoStep`?

nmalfroy replied:
No, because it doesn't get committed until the very end.  The only thing we need to revert explicitly here is the firestore stuff (which we're doing in the `ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep` step

nmalfroy replied:
Eh, you know what, in order to not have that data orphaned, I'll explicitly roll it back

nmalfroy replied:
Gah, actually, I won't.  I'll explicitly rollback any failures since there are a few flight steps involved and that doesn't work for undos

okotsopoulos commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsBqDropStageTableStep.java:
```diff
@@ -0,0 +1,43 @@
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    return StepResult.getStepResultSuccess();
```
Checking that we don't want to restore the staging table in an undo, because if we're unwinding a flight we'll drop the table in the step that creates the table, right?

nmalfroy replied:
that's exactly right!

okotsopoulos reviewed (commented):
This looks pretty good to me, but I have some questions about the desired behavior when a flight fails partway through and needs to be walked back.  Less critical are a few comments around opportunities to avoid repeated work across steps.

snf2ye commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep.java:
```diff
@@ -0,0 +1,50 @@
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    Dataset dataset = datasetService.retrieve(datasetId);
+    // Reverse the id mapping to revert
+    Map<UUID, UUID> oldToNewMappings =
+        ConvertFileIdUtils.readFlightMappings(context.getWorkingMap()).entrySet().stream()
+            .collect(Collectors.toMap(Entry::getValue, Entry::getKey));
```
nit: Call this variable newToOldMappings or mappingsReversed

nmalfroy replied:
gah, good call!

snf2ye commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsBqCreateStageTableStep.java:
```diff
@@ -0,0 +1,64 @@
+    Map<UUID, UUID> oldToNewMappings =
+        ConvertFileIdUtils.readFlightMappings(context.getWorkingMap());
+
+    if (oldToNewMappings.isEmpty()) {
+      logger.info("No file ids to migrate");
+      return StepResult.getStepResultSuccess();
+    }
```
I think this check might be missing for some of the other steps (e.g. ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep) I would suggest moving this check to the flight level so we could skip the steps altogether, but since this is just migration code, not really worth it. 

There are a few places where we don't even end up using the oldToNewMappings and only check if there are any files to migrate. We could add a separate flight key that is just a boolean indicating where the map is empty.

nmalfroy replied:
Yeah, the problem is that we don't know that it's empty until we're in step so we can't read it from the flight.  I only did the check where there the map being empty only causes a single thing to run or not (e.g. in `ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep` we don't need the check...it just iterates on an empty collection)

snf2ye commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsUpdateMetadataStep.java:
```diff
@@ -0,0 +1,26 @@
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    datasetService.setPredictableFileIds(datasetId, true);
+    return StepResult.getStepResultSuccess();
+  }
```
Should there be an undo step here?

nmalfroy replied:
We don't need to worry about it since it never gets committed on failure

nmalfroy replied:
Oh sorry, I this isn't bigquery, this is for Postgres. I didn't want to set it to false automatically (since it could have been true at the start).  Though as I'm typing, I'm realizing that we can just track it at the beginning.  I'll make that change

nmalfroy committed:
PR Feedback cleanup

- Test code cleanup
- Convert sql template creation to set parameters in a fluent style in BigQueryPDao's
- Convert a bunch of steps to use DefaultUndoStep
- Variable renames

nmalfroy committed:
Properly handle rolling back pg metadata if the flight fails

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsGetIdsStep.java:
```diff
@@ -0,0 +1,75 @@
+
+    // Retrieve existing file ids
+    List<String> existingFileIds = fileDao.retrieveAllFileIds(dataset, false);
+    List<FSFile> fsFiles = new ArrayList<>(existingFileIds.size());
+    for (var fileIdBatch :
+        ListUtils.partition(existingFileIds, FireStoreUtils.MAX_FIRESTORE_BATCH_SIZE)) {
+      fsFiles.addAll(fileDao.batchRetrieveById(dataset, fileIdBatch, 0));
+    }
```
This can be done with streaming by using `flatMap()`.
```suggestion
    List<FSFile> fsFiles = ListUtils.partition(existingFileIds, FireStoreUtils.MAX_FIRESTORE_BATCH_SIZE)
        .stream()
        .flatMap(fileIdBatch -> fileDao.batchRetrieveById(dataset, fileIdBatch, 0).stream())
        .toList();
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep implements Step {
+
+  private final UUID datasetId;
+  private final DatasetService datasetService;
+  private final FireStoreDao fileDao;
+
+  public ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep(
+      UUID datasetId, DatasetService datasetService, FireStoreDao fileDao) {
+    this.datasetId = datasetId;
+    this.datasetService = datasetService;
+    this.fileDao = fileDao;
+  }
```
It looks like most `Step`s have all final fields. This means they could be defined using `record` instead of `class`. The main benefit is that you no longer need to write a constructor in addition to the field definitions, since record provides a constructor. In this case, you could write:
```suggestion
public record ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep(UUID datasetId, DatasetService datasetService, FireStoreDao fileDao) implements Step {
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep.java:
```diff
@@ -0,0 +1,50 @@
+    // Reverse the id mapping to revert
+    Map<UUID, UUID> oldToNewMappings =
+        ConvertFileIdUtils.readFlightMappings(context.getWorkingMap()).entrySet().stream()
+            .collect(Collectors.toMap(Entry::getValue, Entry::getKey));
```
Is it guaranteed that all values are unique, e.g., that you never have two different file IDs that both point at the same predictable ID? If there are collisions then this will only reverse one of the move operations.

nmalfroy replied:
Yeah, the idea is that the ids are all 1:1

pshapiro4broad replied:
So you couldn't have two files with different file IDs but with identical hashes? I thought that the original file ID mechanism was random UUID, and the new one is hashcode. So unless we were already mapping identically hashing files to the same random UUID, it seems like you could end up with different "unpredictable" file IDs that have the same predictable file ID.

nmalfroy replied:
Ah so the file id mechanism isn't _just_ the hash.  It's also the size and the target path.  So you can (and often do) have the same file but with a different path.  We treat that as a different file
